### PR TITLE
add TCD Scheduler

### PR DESCRIPTION
**Pull request by mhh0318**
# What does this PR do?

Add a new scheduler for trajectory consistency distillation: [Paper](https://arxiv.org/abs/2402.19159) and [Official Implement](https://github.com/jabir-zheng/TCD/blob/main/scheduling_tcd.py). 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/diffusers/blob/main/CONTRIBUTING.md)?
- [x] Did you read our [philosophy doc](https://github.com/huggingface/diffusers/blob/main/PHILOSOPHY.md) (important for complex PRs)?
- [ ] Was this discussed/approved via a GitHub issue or the [forum](https://discuss.huggingface.co/c/discussion-related-to-httpsgithubcomhuggingfacediffusers/63)? Please add a link to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/diffusers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/diffusers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by yiyixuxu** [comment]
hi!
thanks for your PR :)
does this scheduler with SD and SDXL? can we see some examples?

**Comment by mhh0318** [comment]
Yes, it suits with SD and XL. Currently we only release SDXL LoRA. Here is a demo Space: https://huggingface.co/spaces/h1t/TCD , 
And one example can be:

```
import torch
from diffusers import StableDiffusionXLPipeline

+ from diffusers import TCDScheduler 

device = "cuda"
base_model_id = "stabilityai/stable-diffusion-xl-base-1.0"
tcd_lora_id = "h1t/TCD-SDXL-LoRA"

pipe = StableDiffusionXLPipeline.from_pretrained(base_model_id, torch_dtype=torch.float16, variant="fp16").to(device)
pipe.scheduler = TCDScheduler.from_config(pipe.scheduler.config)

pipe.load_lora_weights(tcd_lora_id)
pipe.fuse_lora()

prompt = "Beautiful woman, bubblegum pink, lemon yellow, minty blue, futuristic, high-detail, epic composition, watercolor."

image = pipe(
    prompt=prompt,
    num_inference_steps=4,
    guidance_scale=0,
    # Eta (referred to as `gamma` in the paper) is used to control the stochasticity in every step.
    # A value of 0.3 often yields good results.
    # We recommend using a higher eta when increasing the number of inference steps.
    eta=0.3, 
    generator=torch.Generator(device=device).manual_seed(0),
).images[0]
```

**Comment by yiyixuxu** [comment]
is this ready for a review?

**Comment by mhh0318** [comment]
Yup. at least looks good for me now.

File: src/diffusers/schedulers/scheduling_tcd.py [hunk 1]
```diff
@@ -0,0 +1,655 @@
...
+        """
+        # 0. Check inputs
+        if num_inference_steps is None and timesteps is None:
+            raise ValueError("Must pass exactly one of `num_inference_steps` or `custom_timesteps`.")
+
+        if num_inference_steps is not None and timesteps is not None:
+            raise ValueError("Can only pass one of `num_inference_steps` or `custom_timesteps`.")
+
+        # 1. Calculate the TCD original training/distillation timestep schedule.
+        original_steps = (
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_tcd.py:368]: it seems like we are handling the case when `original_steps = None` 

but this code here will default `original_steps` to `config.original_inference_steps` when user pass the `None` value - is this intended?

File: src/diffusers/schedulers/scheduling_tcd.py [hunk 2]
```diff
@@ -0,0 +1,655 @@
...
+        while len(sqrt_one_minus_alpha_prod.shape) < len(sample.shape):
+            sqrt_one_minus_alpha_prod = sqrt_one_minus_alpha_prod.unsqueeze(-1)
+
+        velocity = sqrt_alpha_prod * noise - sqrt_one_minus_alpha_prod * sample
+        return velocity
+
+    def __len__(self):
+        return self.config.num_train_timesteps
+
+    def previous_timestep(self, timestep):
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_tcd.py:642]: when do we need this?

File: tests/schedulers/test_scheduler_tcd.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        timesteps = scheduler.timesteps
+        for t in timesteps:
+            self.check_over_forward(time_step=t)
+
+    def test_inference_steps(self):
+        # Hardcoded for now
+        for t, num_inference_steps in zip([99, 39, 39, 19], [10, 25, 26, 50]):
+            self.check_over_forward(time_step=t, num_inference_steps=num_inference_steps)
+
+    # Override test_add_noise_device because the hardcoded num_inference_steps of 100 doesn't work
```
Comment (yiyixuxu) [inline tests/schedulers/test_scheduler_tcd.py:90]: ohh I think you can just override  the `default_num_inference_steps` property, n need to override the test https://github.com/huggingface/diffusers/blob/b9e1c30d0e314b563c1b7ed59c520ebf743dec9f/tests/schedulers/test_schedulers.py#L712

```python
 @property
    def default_num_inference_steps(self):
        return 10
```

File: tests/schedulers/test_scheduler_tcd.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+            sample = self.dummy_sample.to(torch_device)
+            scaled_sample = scheduler.scale_model_input(sample, 0.0)
+            self.assertEqual(sample.shape, scaled_sample.shape)
+
+            noise = torch.randn_like(scaled_sample).to(torch_device)
+            t = scheduler.timesteps[5][None]
+            noised = scheduler.add_noise(scaled_sample, noise, t)
+            self.assertEqual(noised.shape, scaled_sample.shape)
+
+    # Override test_from_save_pretrained because it hardcodes a timestep of 1
```
Comment (yiyixuxu) [inline tests/schedulers/test_scheduler_tcd.py:107]: I think maybe it will work out of the box now with this PR https://github.com/huggingface/diffusers/pull/5479

can you look into if you still need to override this test?

File: tests/schedulers/test_scheduler_tcd.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+            kwargs["generator"] = torch.manual_seed(0)
+            output = scheduler.step(residual, timestep, sample, **kwargs).prev_sample
+
+            kwargs["generator"] = torch.manual_seed(0)
+            new_output = new_scheduler.step(residual, timestep, sample, **kwargs).prev_sample
+
+            assert torch.sum(torch.abs(output - new_output)) < 1e-5, "Scheduler outputs are not identical"
+
+    # Override test_step_shape because uses 0 and 1 as hardcoded timesteps
+    def test_step_shape(self):
```
Comment (yiyixuxu) [inline tests/schedulers/test_scheduler_tcd.py:137]: can you check if you still need to override this test?
with this PR https://github.com/huggingface/diffusers/pull/5479

**Review by yiyixuxu — COMMENTED** [review]
hey thank you for the PR!
looks great and I think we can merge this soon 

I left a few nits,
additionally, we introduced `begin_index` in PR very recently https://github.com/huggingface/diffusers/commit/adcbe674a49af31415a1131174143d58c0e68d7d#diff-98c53466d792a8faefb0045d29513a1c4bcd6b8144b8c06f66b697d4117039d7 - can you apply the same changes here? this will ensure your scheduler working properly with img2img pipelines

File: src/diffusers/schedulers/scheduling_tcd.py (hunk 1 above)
Comment (mhh0318) [reply src/diffusers/schedulers/scheduling_tcd.py:368]: This is an option for setting the range of t from a) ~ [0, 999] or b)~[9, 19, ... 999] whenconfig.original_inference_steps=50.

File: src/diffusers/schedulers/scheduling_tcd.py (hunk 2 above)
Comment (mhh0318) [reply src/diffusers/schedulers/scheduling_tcd.py:642]: In current application we do not need this function. But I found it appears in most of the existing schedulers. And it is used for calculating the previous t, especially we have the t option during set_timesteps.

File: tests/schedulers/test_scheduler_tcd.py (hunk 5 above)
Comment (mhh0318) [reply tests/schedulers/test_scheduler_tcd.py]: removed.

File: tests/schedulers/test_scheduler_tcd.py (hunk 3 above)
Comment (mhh0318) [reply tests/schedulers/test_scheduler_tcd.py]: override `default_num_inference_steps` property and remove following override function

File: src/diffusers/schedulers/scheduling_tcd.py [hunk 6]
```diff
@@ -0,0 +1,655 @@
...
+                The input sample.
+            timestep (`int`, *optional*):
+                The current timestep in the diffusion chain.
+        Returns:
+            `torch.FloatTensor`:
+                A scaled input sample.
+        """
+        return sample
+
+    def _get_variance(self, timestep, prev_timestep):
```
[the remark this answers is not in the document]
Comment (mhh0318) [reply src/diffusers/schedulers/scheduling_tcd.py:319]: added

File: src/diffusers/schedulers/scheduling_tcd.py [hunk 7]
```diff
@@ -0,0 +1,655 @@
...
+        alpha_prod_t = self.alphas_cumprod[timestep]
+        alpha_prod_t_prev = self.alphas_cumprod[prev_timestep] if prev_timestep >= 0 else self.final_alpha_cumprod
+        beta_prod_t = 1 - alpha_prod_t
+        beta_prod_t_prev = 1 - alpha_prod_t_prev
+
+        variance = (beta_prod_t_prev / beta_prod_t) * (1 - alpha_prod_t / alpha_prod_t_prev)
+
+        return variance
+
+    def _threshold_sample(self, sample: torch.FloatTensor) -> torch.FloatTensor:
```
[the remark this answers is not in the document]
Comment (mhh0318) [reply src/diffusers/schedulers/scheduling_tcd.py:330]: added

File: src/diffusers/schedulers/scheduling_tcd.py [hunk 8]
```diff
@@ -0,0 +1,655 @@
...
+
+    betas = []
+    for i in range(num_diffusion_timesteps):
+        t1 = i / num_diffusion_timesteps
+        t2 = (i + 1) / num_diffusion_timesteps
+        betas.append(min(1 - alpha_bar_fn(t2) / alpha_bar_fn(t1), max_beta))
+    return torch.tensor(betas, dtype=torch.float32)
+
+
+def rescale_zero_terminal_snr(betas: torch.FloatTensor) -> torch.FloatTensor:
```
[the remark this answers is not in the document]
Comment (mhh0318) [reply src/diffusers/schedulers/scheduling_tcd.py:97]: added

File: src/diffusers/schedulers/scheduling_tcd.py [hunk 9]
```diff
@@ -0,0 +1,655 @@
...
+            Computed sample `(x_{t-1})` of previous timestep. `prev_sample` should be used as next model input in the
+            denoising loop.
+        pred_noised_sample (`torch.FloatTensor` of shape `(batch_size, num_channels, height, width)` for images):
+            The predicted noised sample `(x_{s})` based on the model output from the current timestep.
+    """
+
+    prev_sample: torch.FloatTensor
+    pred_noised_sample: Optional[torch.FloatTensor] = None
+
+
```
[the remark this answers is not in the document]
Comment (mhh0318) [reply src/diffusers/schedulers/scheduling_tcd.py:50]: added

File: src/diffusers/schedulers/scheduling_tcd.py (hunk 1 above)
Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_tcd.py:398]: ok, got it 
so I think there might be an error in the below code then, if user pass `original_inference_steps = None`, `orignal_steps` will be set to `config.original_inference_steps` , so `orignal_steps` won't be None and the else part won't get evaluated 

```python
        if original_steps is not None:
            if original_steps > self.config.num_train_timesteps:
                raise ValueError(
                    f"`original_steps`: {original_steps} cannot be larger than `self.config.train_timesteps`:"
                    f" {self.config.num_train_timesteps} as the unet model trained with this scheduler can only handle"
                    f" maximal {self.config.num_train_timesteps} timesteps."
                )
            # TCD Timesteps Setting
            # The skipping step parameter k from the paper.
            k = self.config.num_train_timesteps // original_steps
            # TCD Training/Distillation Steps Schedule
            tcd_origin_timesteps = np.asarray(list(range(1, int(original_steps * strength) + 1))) * k - 1
        else:
            tcd_origin_timesteps = np.asarray(list(range(0, int(self.config.num_train_timesteps * strength))))
```

File: src/diffusers/schedulers/scheduling_tcd.py (hunk 2 above)
Comment (yiyixuxu) [reply src/diffusers/schedulers/scheduling_tcd.py:673]: ok!

**Review by yiyixuxu — APPROVED** [review]
left one question
let me know, if it is not an error

PR looks ready to merge for me. thanks for the great work!

File: src/diffusers/schedulers/scheduling_tcd.py [hunk 10]
```diff
@@ -0,0 +1,686 @@
...
+        if num_inference_steps is not None and timesteps is not None:
+            raise ValueError("Can only pass one of `num_inference_steps` or `custom_timesteps`.")
+
+        # 1. Calculate the TCD original training/distillation timestep schedule.
+        original_steps = (
+            original_inference_steps if original_inference_steps is not None else self.config.original_inference_steps
+        )
+
+        if original_steps is not None:
```
Comment (yiyixuxu) [inline src/diffusers/schedulers/scheduling_tcd.py:402]:
```suggestion
        if original_inference_steps is not None:
```

 maybe this? not sure.. 
see https://github.com/huggingface/diffusers/pull/7174/files#r1511499565

**Review by sayakpaul — COMMENTED** [review]
Nice!

I think there also needs to a guide in our docs for the users to know how to best use TCD just like we have https://huggingface.co/docs/diffusers/main/en/using-diffusers/inference_with_lcm_lora. 

@yiyixuxu WDYT?

File: src/diffusers/schedulers/scheduling_tcd.py (hunk 10 above)
Comment (mhh0318) [reply src/diffusers/schedulers/scheduling_tcd.py:402]: `else` should only be evaluated with customised training steps or some special situation. I think following would be better:
```
# default option
if original_inference_steps is None: 
  tcd_origin_timesteps align with discrete inference steps
# customised 
else:
  tcd_origin_timesteps truly randomised
```

**Comment by yiyixuxu** [comment]
let us know when it's ready to merge 

>I think there also needs to a guide in our docs for the users to know how to best use TCD just like we have https://huggingface.co/docs/diffusers/main/en/using-diffusers/inference_with_lcm_lora.

what @sayakpaul suggested here is a great way to help promote this scheduler, feel free to add a guide either in this PR or a separate one

**Comment by mhh0318** [comment]
> let us know when it's ready to merge
> 
> > I think there also needs to a guide in our docs for the users to know how to best use TCD just like we have https://huggingface.co/docs/diffusers/main/en/using-diffusers/inference_with_lcm_lora.
> 
> what @sayakpaul suggested here is a great way to help promote this scheduler, feel free to add a guide either in this PR or a separate one

I think it is ready to merge for TCD scheduler now. I would open a new PR for the guideline.
